dynamic_programming/levenshtein_distance.nim

```diff
@@ -0,0 +1,72 @@
+## Levenshtein distance
+
+import tables
+
+
+func initSubsolutions(lenA: Natural, lenB: Natural): Table[(Natural, Natural), Natural] =
+  for indA in 0..lenA:
+    result[(Natural(indA), Natural(0))] = Natural(indA)
+
+  for indB in 1..lenB:
+    result[(Natural(0), Natural(indB))] = Natural(indB)
+
+
+func toKey(indA: int, indB: int): (Natural, Natural) =
+  return (Natural(indA), Natural(indB))
+
+
+func computeLevenshteinDistanceMatrix(strA: string, strB: string): Table[(
+    Natural, Natural), Natural] =
+  let
+    lenA = strA.len
+    lenB = strB.len
+
+  result = initSubsolutions(lenA, lenB)
+
+  for indA in 1..lenA:
+    for indB in 1..lenB:
+      let substitutionCost = (if strA[indA-1] == strB[indB-1]: 0 else: 1)
+      result[toKey(indA, indB)] = min(
+          [result[toKey(indA-1, indB)] + 1,
+           result[toKey(indA, indB-1)] + 1,
+           result[toKey(indA-1, indB-1)] + substitutionCost])
```

Comment:
```suggestion
func computeLevenshteinDistanceMatrix(a, b: string): Table[(Natural, Natural),
                                                           Natural] =
  result = initSubsolutions(a.len, b.len)

  for indA in 0..<a.len:
    for indB in 0..<b.len:
      let substitutionCost = (if a[indA] == b[indB]: 0 else: 1)
      result[toKey(indA + 1, indB + 1)] = min(
          [result[toKey(indA, indB + 1)] + 1,
           result[toKey(indA + 1, indB)] + 1,
           result[toKey(indA, indB)] + substitutionCost])
```
Even though our guidelines [call for restraining from single-letter variables](/TheAlgorithms/Nim/CONTRIBUTING.md#readability-and-naming-conventions), it's the perfect example where they are suitable and non-ambiguous. Both arguments of the function have the same type and are commutative, and their type is just at 6-line distance away tops.

Also, if an algorithm has an exact analogue but with addition and not subtraction, I prefer the former. Don't have any serious arguments besides the fact that people tend to reason better about addition.

You probably know this, but just to clarify, `string.len` is a field and not a function, so storing it in a variable is not an optimisation and is a matter of readability.